rl4co/envs/routing/mpdp/env.py

```diff
@@ -131,32 +138,79 @@ def _step(self, td: TensorDict) -> TensorDict:
+    @staticmethod
+    def get_action_mask(td: TensorDict) -> torch.Tensor:
+        """Get the action mask for the current state."""
+
+        visited_loc = td["visited"].clone()
+
+        agent_num = td["lengths"].size(1)
+        n_loc = visited_loc.size(-1) - agent_num - 1  # num of customers
+        batch_size = visited_loc.size(0)
+        agent_idx = td["agent_idx"][:, None, :]
+        mask_loc = visited_loc.to(td["to_delivery"].device) | (1 - td["to_delivery"])
+
+        # depot
+        if td["i"][0].item() != 0:
+            mask_loc[:, :, : agent_num + 1] = 1
 
-        if td is None or td.is_empty():
-            td = self.generate_data(batch_size=batch_size)
+            # if deliver nodes which is assigned agent is complete, then agent can go to depot
+            no_item_to_delivery = (
+                visited_loc[:, :, n_loc // 2 + agent_num + 1 :]
+                == td["to_delivery"][:, :, n_loc // 2 + agent_num + 1 :]
+            ).all(dim=-1)
+            mask_loc[no_item_to_delivery.squeeze(-1), :, :] = mask_loc[
+                no_item_to_delivery.squeeze(-1), :, :
+            ].scatter_(-1, agent_idx[no_item_to_delivery.squeeze(-1), :, :], 0)
+
+            condition = (td["count_depot"] == agent_num - 1) & (
+                (visited_loc[:, :, agent_num + 1 :] == 0).sum(dim=-1) != 0
+            )
 
-        self.to(td.device)
+            mask_loc[..., agent_num][condition] = 1
+
+        else:
+            return (
+                torch.cat(
+                    [
+                        torch.zeros(
+                            batch_size, 1, 1, dtype=torch.uint8, device=mask_loc.device
+                        ),
+                        torch.ones(
+                            batch_size,
+                            1,
+                            n_loc + agent_num,
+                            dtype=torch.uint8,
+                            device=mask_loc.device,
+                        ),
+                    ],
+                    dim=-1,
+                )
+                > 0
+            )
+        action_mask = mask_loc == 0  # action_mask gets feasible actions
+        return action_mask
+
+    def _reset(self, td: Optional[TensorDict] = None, batch_size: Optional[list] = None) -> TensorDict:
+        device = td.device
+        locs = td["locs"]
+        num_loc = locs.size(-2)
+        
+        visited = torch.zeros((*batch_size, num_loc), dtype=torch.bool, device=device)
+
+        # Depot is always visited
+        visited[:, 0] = True
 
         # NOTE: this is a hack to get the agent_num
-        # agent_num = td["agent_num"][0].item() if agent_num is None else agent_num
-        # agent_num = agent_num if agent_num is not None else td["agent_num"][0].item()
+        agent_num = td["agent_num"][0].item() if agent_num is None else agent_num
 
         depot = td["depot"]
         depot = depot.repeat(1, agent_num + 1, 1)
         loc = td["locs"]
```

Comment:
You're setting `locs = td["locs"]` in line 196 and `loc = td["locs"]` in line 209. We could use one and replace the occurences of the other in the code (below).